File: java/Decoupling-Temporal-Services-with-Nexus/exercise/src/main/java/compliance/temporal/workflow/ComplianceWorkflowImpl.java
package compliance.temporal.workflow;

// ═══════════════════════════════════════════════════════════════════
//  TODO 2: Implement the ComplianceWorkflow
// ═══════════════════════════════════════════════════════════════════
//
// This workflow:
//   1. Runs ComplianceActivity.checkCompliance() on the request
//   2. If LOW or HIGH risk → returns immediately
//   3. If MEDIUM risk → waits for a human review() Update
//
// ── What to implement: ──────────────────────────────────────────
//
//   Fields:
//     - ComplianceRequest request (store the input)
//     - ComplianceResult autoResult (store the automated check result)
//     - ComplianceResult reviewResult = null (set by review Update)
//     - An activity stub for ComplianceActivity (30s startToCloseTimeout)
//
//   run(ComplianceRequest request):
//     1. Store this.request = request
//     2. Call complianceActivity.checkCompliance(request) → autoResult
//     3. Add Workflow.sleep(Duration.ofSeconds(10)) to simulate slow processing
//        (durable: kill the compliance worker mid-sleep, restart it — the workflow resumes)
//     4. If autoResult risk is NOT "MEDIUM" → return autoResult
//     5. Otherwise: Workflow.await(() -> reviewResult != null)
//     6. Return reviewResult
//
//   review(boolean approved, String explanation):
//     - Create a new ComplianceResult with the review decision
//     - Store it in reviewResult (this unblocks run())
//     - Return reviewResult
//
//   validateReview(boolean approved, String explanation):
//     - Throw IllegalStateException if not awaiting review
//     - Throw IllegalStateException if review already submitted

import compliance.domain.ComplianceRequest;
import compliance.domain.ComplianceResult;
import compliance.temporal.activity.ComplianceActivity;
import io.temporal.activity.ActivityOptions;
import io.temporal.workflow.Workflow;

import java.time.Duration;

public class ComplianceWorkflowImpl implements ComplianceWorkflow {

    private ComplianceRequest request;
    private ComplianceResult autoResult;
    private ComplianceResult reviewResult = null;

    private final ComplianceActivity complianceActivity = Workflow.newActivityStub(
            ComplianceActivity.class,
            ActivityOptions.newBuilder()
                    .setStartToCloseTimeout(Duration.ofSeconds(30))
                    .build());

    @Override
    public ComplianceResult run(ComplianceRequest request) {
        // TODO: Store the request, run the activity, add Workflow.sleep(10s),
        //       check risk level: if MEDIUM → Workflow.await(() -> reviewResult != null)
        //       otherwise → return autoResult directly
        return null;
    }

    @Override
    public ComplianceResult review(boolean approved, String explanation) {
        // TODO: Create ComplianceResult from the review decision, store in reviewResult
        return null;
    }

    @Override
    public void validateReview(boolean approved, String explanation) {
        // TODO: Throw IllegalStateException if not awaiting review or already reviewed
    }
}

File: kotlin/decouple-monolith/solution/src/main/kotlin/compliance/temporal/workflow/ComplianceWorkflowImpl.kt
package compliance.temporal.workflow

import compliance.domain.ComplianceRequest
import compliance.domain.ComplianceResult
import compliance.temporal.activity.ComplianceActivity
import io.temporal.activity.ActivityOptions
import io.temporal.workflow.Workflow
import java.time.Duration

/**
 * [GIVEN] Compliance Workflow implementation. Pre-built for this exercise.
 *
 * The 10-second Workflow.sleep is intentional. It gives you a window to stop the
 * Compliance Worker with Ctrl+C and restart it, so you can watch the Nexus Operation
 * survive the outage. The timer runs on the server, so nothing is lost when the
 * Worker goes away.
 */
class ComplianceWorkflowImpl : ComplianceWorkflow {

    private var request: ComplianceRequest? = null
    private var autoResult: ComplianceResult? = null
    private var reviewResult: ComplianceResult? = null

    private val complianceActivity: ComplianceActivity = Workflow.newActivityStub(
        ComplianceActivity::class.java,
        ActivityOptions.newBuilder()
            .setStartToCloseTimeout(Duration.ofSeconds(30))
            .build(),
    )

    override fun run(request: ComplianceRequest): ComplianceResult {
        this.request = request

        // Step 1: Run the automated compliance check
        val auto = complianceActivity.checkCompliance(request)
        autoResult = auto

        // Durable delay. Stop the Compliance Worker mid-sleep, restart it, and the
        // Workflow resumes on its own.
        Workflow.sleep(Duration.ofSeconds(10))

        // Step 2: LOW or HIGH risk returns immediately
        if (auto.riskLevel != "MEDIUM") {
            return auto
        }

        // Step 3: MEDIUM risk waits for human review via Update.
        // Bounded on purpose. An unbounded await would leave this Workflow Running
        // forever when nobody reviews. Eight minutes is deliberately shorter than the
        // caller's 10-minute scheduleToCloseTimeout, so the handler resolves and reports
        // a decision rather than letting the caller time out first.
        val reviewed = Workflow.await(Duration.ofMinutes(8)) { reviewResult != null }
        if (!reviewed) {
            return ComplianceResult(
                transactionId = request.transactionId,
                approved = false,
                riskLevel = "MEDIUM",
                explanation = "No reviewer responded within 8 minutes. Not approved.",
            )
        }
        return reviewResult!!
    }

    override fun review(approved: Boolean, explanation: String): ComplianceResult {
        val result = ComplianceResult(
            transactionId = request!!.transactionId,
            approved = approved,
            riskLevel = "MEDIUM",
            explanation = explanation,
        )
        reviewResult = result
        return result
    }

    override fun validateReview(approved: Boolean, explanation: String) {
        check(autoResult?.riskLevel == "MEDIUM") { "Workflow is not awaiting review" }
        check(reviewResult == null) { "Review already submitted" }
    }
}
